src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,617 @@
+        # If the device does not have an attribute of data type octet string, skip this step.
+
+        # Verify on the TH that the DUT returns data from the expected attribute path.
+        self.print_step(14, "Read a read request of attribute type octet string")
+        await self.get_cluster_from_type(bytes)
+
+        # Step 15
+        # TH sends the Read Request Message to the DUT to read an attribute of data type Struct.
+        # If the device does not have an attribute of data type struct, skip this step.
+
+        # Verify on the TH that the DUT returns data from the expected attribute path.
+        self.print_step(15, "Read a read request of attribute type struct")
+        await self.get_cluster_from_type(ClusterObject)
+
+        # Step 16
+        # TH sends the Read Request Message to the DUT to read an attribute of data type List.
+        # If the device does not have an attribute of data type list, skip this step.
+        # Verify on the TH that the DUT returns data from the expected attribute path.
+        self.print_step(16, "Read a read request of attribute type list")
+        await self.get_cluster_from_type(list)
+
+        # Step 17
+        # TH sends the Read Request Message to the DUT to read an attribute of data type enum.
+        # If the device does not have an attribute of data type enum, skip this step.
+        # Verify on the TH that the DUT returns data from the expected attribute path.
+        self.print_step(17, "Read a read request of attribute type enum")
+        await self.get_cluster_from_type(MatterIntEnum)
+
+        # Step 18
+        # TH sends the Read Request Message to the DUT to read an attribute of data type bitmap.
+        # If the device does not have an attribute of data type bitmap, skip this step.
+        self.print_step(18, "Read a read request of attribute type bitmap")
+        await self.get_cluster_from_type(IntFlag)
+
+        # Step 19
+        # TH sends the Read Request Message to the DUT to read any attribute to an unsupported Endpoint.
+        # DUT responds with the report data action.
+        # Verify on the TH that the DUT sends the status code UNSUPPORTED_ENDPOINT
+
+        self.print_step(19, "Send the Read Request Message to the DUT to read any attribute to an unsupported Endpoint")
+        supported_endpoints = set(self.endpoints.keys())
+        all_endpoints = set(range(max(supported_endpoints)+2))
+        unsupported = list(all_endpoints - supported_endpoints)
+        # Read descriptor
+        result = await self.read_single_attribute_expect_error(endpoint=unsupported[0], cluster=Clusters.Descriptor, attribute=Clusters.Descriptor.Attributes.FeatureMap, error=Status.UnsupportedEndpoint)
+        asserts.assert_true(isinstance(result.Reason, InteractionModelError), msg="Unexpected success reading invalid endpoint")
+
+        # Seems to return only {} -- Is this a failure or the unintended behavior?
```

Comment:
what returns {}?

endpoint should be in a tuple - [(non_existent_endpoint)]

Reply:
result returns a tuple in this case, but I can fix the endpoint to make it work correctly.